File: teste_threds.py
import multiprocessing
import time
import random
import threading

def pico_de_estresse():
    """Simula um pico de estresse aleatório a cada 5 segundos em loop infinito."""
    while True:
        time.sleep(5)  # Espera 5 segundos
        print("Pico de estresse iniciado!")
        
        # Criando um "pico de estresse" aumentando a carga por um tempo
        duracao_pico = random.uniform(0.5, 2.0)  # Duração aleatória do pico de estresse
        print(f"Pico de estresse durando {duracao_pico:.2f} segundos")
        # Simula o aumento de carga de CPU por essa duração
        start = time.time()
        while time.time() - start < duracao_pico:
            pass  # Realiza uma operação de "busy-wait" para estressar a CPU

        print("Pico de estresse finalizado!")

def tarefa_pesada(n):
    while True:  # Loop infinito para as tarefas pesadas
        print(f"Iniciando tarefa pesada {n}")
        time.sleep(random.uniform(1, 3))  # Simula uma tarefa pesada
        print(f"Finalizando tarefa pesada {n}")

def tarefa_leve(n):
    while True:  # Loop infinito para as tarefas leves
        print(f"Iniciando tarefa leve {n}")
        time.sleep(random.uniform(0.1, 0.5))  # Simula uma tarefa leve
        print(f"Finalizando tarefa leve {n}")

def main():
    num_nucleos = multiprocessing.cpu_count()  # Descobre o número de núcleos disponíveis
    tarefas_pesadas = [multiprocessing.Process(target=tarefa_pesada, args=(i,)) for i in range(10)]
    tarefas_leves = [multiprocessing.Process(target=tarefa_leve, args=(i,)) for i in range(6)]
    
    # Inicia o pico de estresse em um thread separado
    stress_thread = threading.Thread(target=pico_de_estresse, daemon=False)
    stress_thread.start()

    # Inicia todas as tarefas pesadas (usando 4 núcleos, por exemplo)
    for i in range(4):  # Distribui as 4 primeiras tarefas pesadas nos núcleos
        tarefas_pesadas[i].start()

    # Inicia as tarefas leves, aproveitando os núcleos ociosos
    for i in range(6):  # Começa as tarefas leves
        tarefas_leves[i].start()

    # Espera as tarefas pesadas restantes terminarem (na verdade, elas vão rodar infinitamente)
    for i in range(4, 10):
        tarefas_pesadas[i].start()

    # Não chamamos join() para as tarefas, pois elas agora são infinitas.
    # O processo principal continuará rodando infinitamente até que o programa seja interrompido manualmente.
    while True:
        time.sleep(1)  # Manter o loop do processo principal ativo.

def pico_de_estresse():

if __name__ == "__main__":
    main()

File: teste_threds.java
import java.util.Random;

public class TesteStress {

    public static void main(String[] args) {
        // Cria e inicia o pico de estresse em uma nova thread
        Thread stressThread = new Thread(new PicoDeEstresse());
        stressThread.setDaemon(false); // A thread de pico de estresse continuará mesmo após o término do main
        stressThread.start();

        // Cria e inicia as tarefas pesadas
        for (int i = 0; i < 10; i++) {
            final int n = i;
            Thread tarefaPesada = new Thread(() -> tarefaPesada(n));
            tarefaPesada.setDaemon(true); // As tarefas pesadas rodarão em threads separadas
            tarefaPesada.start();
        }

        // Cria e inicia as tarefas leves
        for (int i = 0; i < 6; i++) {
            final int n = i;
            Thread tarefaLeve = new Thread(() -> tarefaLeve(n));
            tarefaLeve.setDaemon(true); // As tarefas leves também em threads separadas
            tarefaLeve.start();
        }

        // Loop infinito no thread principal para manter o programa em execução
        while (true) {
            try {
                Thread.sleep(1000); // Apenas para evitar o fechamento imediato do main
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    // Função para simular um pico de estresse de CPU
    public static class PicoDeEstresse implements Runnable {
        @Override
        public void run() {
            while (true) {
                try {
                    Thread.sleep(5000); // Espera 5 segundos
                    System.out.println("Pico de estresse iniciado!");

                    // Cria um "pico de estresse" com duração aleatória entre 0.5 e 2 segundos
                    Random rand = new Random();
                    double duracaoPico = 0.5 + (1.5 * rand.nextDouble());
                    System.out.printf("Pico de estresse durando %.2f segundos%n", duracaoPico);

                    // Simula a carga de CPU para o tempo do pico de estresse
                    long startTime = System.currentTimeMillis();
                    while (System.currentTimeMillis() - startTime < duracaoPico * 1000) {
                        // Busy-wait para sobrecarregar a CPU
                    }

                    System.out.println("Pico de estresse finalizado!");
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }
    }

    // Função que simula uma tarefa pesada
    public static void tarefaPesada(int n) {
        while (true) {
            try {
                System.out.printf("Iniciando tarefa pesada %d%n", n);
                Thread.sleep(1000 + new Random().nextInt(2000)); // Simula uma tarefa pesada
                System.out.printf("Finalizando tarefa pesada %d%n", n);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    // Função que simula uma tarefa leve
    public static void tarefaLeve(int n) {
        while (true) {
            try {
                System.out.printf("Iniciando tarefa leve %d%n", n);
                Thread.sleep(100 + new Random().nextInt(400)); // Simula uma tarefa leve
                System.out.printf("Finalizando tarefa leve %d%n", n);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
